probe/awsecs/client.go

```diff
@@ -106,45 +239,166 @@ func (c ecsClient) getTasks(taskArns []string) (map[string]*ecs.Task, error) {
 		Tasks:   taskPtrs,
 	})
 	if err != nil {
-		return nil, err
+		log.Warnf("Failed to describe ECS tasks, ECS service report may be incomplete: %v", err)
+		return
 	}
 
 	for _, failure := range resp.Failures {
 		log.Warnf("Failed to describe ECS task %s, ECS service report may be incomplete: %s", *failure.Arn, *failure.Reason)
 	}
 
-	results := make(map[string]*ecs.Task, len(resp.Tasks))
 	for _, task := range resp.Tasks {
-		results[*task.TaskArn] = task
+		c.taskCache.Set(*task.TaskArn, newECSTask(task))
+	}
+}
+
+// Try to match a list of task ARNs to service names using cached info.
+// Returns (task to service map, unmatched tasks). Ignores tasks whose startedby values
+// don't appear to point to a service.
+func (c ecsClientImpl) matchTasksServices(taskARNs []string) (map[string]string, []string) {
+	deploymentMap := map[string]string{}
+	for _, serviceNameRaw := range c.serviceCache.Keys() {
+		serviceName := serviceNameRaw.(string)
+		service, ok := c.getCachedService(serviceName)
+		if !ok {
+			// This is rare, but possible if service was evicted after the loop began
+			continue
+		}
+		for _, deployment := range service.DeploymentIDs {
+			deploymentMap[deployment] = serviceName
+		}
+	}
+	log.Debugf("Mapped %d deployments from %d services", len(deploymentMap), c.serviceCache.Len())
+
+	results := map[string]string{}
+	unmatched := []string{}
+	for _, taskARN := range taskARNs {
+		task, ok := c.getCachedTask(taskARN)
+		if !ok {
+			// this can happen if we have a failure while describing tasks, just pretend the task doesn't exist
+			continue
+		}
+		if !task.IsServiceManaged() {
+			continue
+		}
+		if serviceName, ok := deploymentMap[task.StartedBy]; ok {
+			results[taskARN] = serviceName
+		} else {
+			unmatched = append(unmatched, taskARN)
+		}
+	}
+
+	log.Debugf("Matched %d from %d tasks, %d unmatched", len(results), len(taskARNs), len(unmatched))
+	return results, unmatched
+}
+
+func (c ecsClientImpl) ensureTasksAreCached(taskARNs []string) {
+	tasksToFetch := []string{}
+	for _, taskARN := range taskARNs {
+		if _, err := c.taskCache.Get(taskARN); err != nil {
+			tasksToFetch = append(tasksToFetch, taskARN)
+		}
 	}
-	return results, nil
+	if len(tasksToFetch) > 0 {
+		// This might not fully succeed, but we only try once and ignore any further missing tasks.
+		c.getTasks(tasksToFetch)
+	}
+}
+
+func (c ecsClientImpl) refreshServices(taskServiceMap map[string]string) map[string]bool {
+	toDescribe, done := c.describeServices()
+	servicesRefreshed := map[string]bool{}
+	for _, serviceName := range taskServiceMap {
+		if servicesRefreshed[serviceName] {
+			continue
+		}
+		toDescribe <- serviceName
+		servicesRefreshed[serviceName] = true
+	}
+	close(toDescribe)
+	<-done
```

Comment:
This is the read here, we're blocking on it being written to or closed inside describeServices. However, by simplifying describeServices to take in args instead of returning an input chan, we will do away with the need for it.